Add command to toggle clean mode

aphymi opened the pull request:
This PR resolves #1500 by adding a built-in editor command to toggle 'clean mode', i.e. the feature that hides markdown syntax like `*`s when not focused by the cursor.

The goal of having clean mode disabled is to have *both* the markdown syntax and formatting simultaneously, except for places where that would be overly repetitive (e.g. tables), in which case the syntax is prioritized.

Implementation-wise, I largely mimicked the `Toggle Dark Mode` command for the toggling behavior, creating a `cleanMode` UI option, defaulting to true. When false, almost every instance of markdown syntax being hidden is disabled, as if everything on the page was alt-clicked or highlighted. The only exception is lua directives (`${...}`); their behavor is unchanged whether clean mode is enabled or not.

There are, I think, two questions left in my mind before I'd call this done:
1. *Should* lua directives be excluded from this? The other reasonable option, I think, is giving them the same behavior as images (and other inline content), rendering both the syntax and the result. I poked at that a bit, but it behaved oddly.
2. Should the color of syntax characters change with clean mode disabled? The current `--meta-color` in the dark theme is a rather angry red, giving the impression that there's some kind of error with syntax characters. It'd be nice if that was less the case with clean mode disabled, where syntax characters are an expected part of the page. This could be changed on a per-user/space basis with space styles, of course.

MrMugame commented on web/editor_ui.tsx:
```diff
@@ -127,6 +127,15 @@ export class MainUI {
     }, [viewState.uiOptions.darkMode]);
 
+    useEffect(() => {
+      clientStoreSyscalls(client.ds)["clientStore.get"](
```
The `darkMode` example above sets a really bad precedent. I don't know what @zefhemel thinks of this, but IMO, this isn't at all how this should be done. (Also the `toggleDarkMode` is bad imo, it gets the option from the ui options and sets it in the clientstore). ~~At least `client.clientSystem.localSyscall` should have been used.~~

The clientstore should be solely used to cache the option in the browser, so it persists across reloads. It should never be touched inside of here and most definitely not by constructing some new syscalls. The code in here should only look at the ui option and update accordingly. The editor plug should set that ui option. (Now this is more of a personal opinion, but I think it should also be `darkMode: "system" | "dark" | "light"`, I think that reflects it better, but that really doesn't matter and I'm unsure if changing that may break some plug code). The `vimMode` option seems to do all of this properly.

This isn't your fault, I just think this should be changed. It may work right now, but that's not really an argument right here.

MrMugame commented on web/cm_plugins/wiki_link.ts:
```diff
@@ -116,7 +123,7 @@ export function cleanWikiLinkPlugin(client: Client) {
           Decoration.replace({
             widget: new LinkWidget({
-              text: linkText,
+              text: cleanModeEnabled ? linkText : textWithoutMarks,
```
No reason to do all of this again I think. Just bail out at the `isCursorInRange` check.

aphymi replied:
...oh! You're right! Thanks for pointing that out!

MrMugame commented on web/ui_types.ts:
```diff
@@ -76,6 +77,7 @@ export const initialViewState: AppViewState = {
   uiOptions: {
     vimMode: false,
     darkMode: undefined,
+    cleanMode: true,
```
This should be `false` by default

aphymi replied:
To be clear: `true` is the *current* behavior, and setting it to `false` triggers the new behavior introduced by this PR. Are you saying you want this new behavior to be default?

MrMugame replied:
Oh, ok, I interpreted `cleanMode` as in "cleaned from any styling". I guess that's a perspective thing.

PowerUser64 commented:
Maybe it would be good to rename this option. _Clean mode_ might be what it's called someplace internally, but I think we might want to come up with a different name to present to users. I was confused when I saw #1500 at first - I thought "ooh, clean mode - what's that?" After seeing similar confusion in @MrMugame's comment [above](https://github.com/silverbulletmd/silverbullet/pull/1502#discussion_r2298012168), I think it's not just me. Perhaps calling it something like "inline rendering" it would be more immediately obvious. Thoughts? Feel free to share other name ideas too if you think of any.

aphymi commented on web/cm_plugins/wiki_link.ts:
```diff
@@ -19,6 +19,8 @@ import { wikiLinkRegex } from "../markdown_parser/constants.ts";
  * Plugin to hide path prefix when the cursor is not inside.
  */
 export function cleanWikiLinkPlugin(client: Client) {
+  const cleanModeEnabled = client.ui.viewState.uiOptions.cleanMode;
```
Hmmm, I agree that this placement is awkward, but I don't think passing it as a parameter would be better, given that the client is already being passed, with all its context. I've moved the definition closer to its only remaining use now, anyway.

aphymi commented:
@PowerUser64 
> Perhaps calling it something like "inline rendering" it would be more immediately obvious. Thoughts? Feel free to share other name ideas too if you think of any.

Yeah, "clean mode" isn't necessarily intuitive; I'd be up to change that. 'inline rendering' feels too nonspecific. Maybe 'Toggle Markdown Syntax Rendering'? It's a bit of a mouthful, but it's probably the clearest option?

@MrMugame
> My guess would be some margin: -1em 0; problems. It's all very finicky.

Oh! Yes, that was exactly it. Don't know how I missed that.

> But overall, that behavior would seem the most intuitive to me.

Fair enough! I've done that in b1d856c. The only catch is that the rendered result still hides while you're editing the directive. Lua blocks seem to have a flash of an incorrect render every refresh before they settle down, but that gets really jittery and ugly when a refresh happens with every keystroke, so I chose to make them hide until your cursor leaves the directive.

MrMugame commented:
> Lua blocks seem to have a flash of an incorrect render every refresh before they settle down, but that gets really jittery and ugly when a refresh happens with every keystroke, so I chose to make them hide until your cursor leaves the directive.

That's something to consider with inlined content (`![]()`) in the current/"normal" mode, because it's facing the exact same problem and this also currently causes a bug with autocomplete not working. That's would be a pretty easy fix. Unsure what @zef thinks.

aphymi commented:
I've renamed the command to `Toggle Markdown Syntax Rendering`, and added a note about it to a couple relevant documentation pages. Absent a fix for that rendering weirdness, I think this feature is ready!
